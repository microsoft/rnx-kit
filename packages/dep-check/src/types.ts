import type { Capability, KitType } from "@rnx-kit/config";
import type { PackageManifest } from "@rnx-kit/tools-node/package";

type Options = {
  customProfiles?: string;
  excludePackages?: string;
  loose: boolean;
  write: boolean;
};

export type Args = Options & {
  "custom-profiles"?: string | number;
  "exclude-packages"?: string | number;
  "set-version"?: string | number;
  init?: string;
  packages?: (string | number)[];
  vigilant?: string | number;
};

export type CheckConfig = {
  kitType: KitType;
  reactNativeVersion: string;
  reactNativeDevVersion: string;
  capabilities: Capability[];
  customProfilesPath?: string;
  manifest: PackageManifest;
};

export type CheckOptions = Options & {
  uncheckedReturnCode?: number;
  config?: number | CheckConfig;
  supportedVersions?: string;
  targetVersion?: string;
};

export type VigilantOptions = Options & {
  versions: string;
};

export type CapabilitiesOptions = {
  kitType?: KitType;
  customProfilesPath?: string;
};

export type Command = (manifest: string) => number;

export type DependencyType = "direct" | "development" | "peer";

export type MetaPackage = {
  name: "#meta";
  capabilities: Capability[];
  devOnly?: boolean;
};

export type Package = {
  name: string;
  version: string;
  capabilities?: Capability[];
  devOnly?: boolean;
};

export type ManifestProfile = PackageManifest & {
  dependencies: Record<string, string>;
  peerDependencies: Record<string, string>;
  devDependencies: Record<string, string>;
};

export type Profile = Readonly<Record<Capability, MetaPackage | Package>>;

export type ProfileVersion =
  | "0.61"
  | "0.62"
  | "0.63"
  | "0.64"
  | "0.65"
  | "0.66"
  | "0.67"
  | "0.68"
  | "0.69"
  | "0.70";

export type ProfileMap = Record<ProfileVersion, Profile>;

export type ProfilesInfo = {
  supportedProfiles: Profile[];
  supportedVersions: string;
  targetProfile: Profile[];
  targetVersion: string;
};

export type ExcludedPackage = Package & {
  reason: string;
};
