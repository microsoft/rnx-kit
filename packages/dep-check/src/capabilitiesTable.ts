import type { Capability } from "@rnx-kit/config";
import { ProfileVersion, getAllProfiles } from "./profiles";

function maxStringLength(currentMax: number, str: string): number {
  return Math.max(currentMax, str.length);
}

function sortCoreFirst(lhs: string, rhs: string): number {
  if (lhs.startsWith("core-")) {
    if (!rhs.startsWith("core-")) {
      return -1;
    }
  } else if (rhs.startsWith("core-")) {
    return 1;
  }

  if (lhs === rhs) {
    return 0;
  }

  return lhs < rhs ? -1 : 1;
}

const allProfiles = getAllProfiles();
const allVersions = Object.keys(allProfiles).reverse() as ProfileVersion[];
const allCapabilities = Object.keys(allProfiles["0.64"]).sort(
  sortCoreFirst
) as Capability[];

const headers = ["Capability", ...allVersions];

const columns = allVersions.reduce<string[][]>(
  (result, version) => {
    const profile = allProfiles[version];
    const column = allCapabilities.map((capability) => {
      const { name, version } = profile[capability];
      return `${name}@${version}`;
    });
    result.push(column);
    return result;
  },
  [allCapabilities]
);

const columnWidths = columns.map((column, index) =>
  column.reduce(maxStringLength, headers[index].length)
);

const pad = (s: string, index: number) => s.padEnd(columnWidths[index]);

const table = [
  `| ${headers.map(pad).join(" | ")} |`,
  `| ${columnWidths.map((width) => "-".repeat(width)).join(" | ")} |`,
  ...allCapabilities.map(
    (_, capability) =>
      `| ${columns
        .map((column, index) => pad(column[capability], index))
        .join(" | ")} |`
  ),
].join("\n");

console.log(table);
