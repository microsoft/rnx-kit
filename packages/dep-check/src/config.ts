import type { KitConfig } from "@rnx-kit/config";
import semver from "semver";
import { getRequirements } from "./dependencies";
import type { PackageManifest } from "./types";

type KitManifest = Required<
  Pick<
    KitConfig,
    "capabilities" | "kitType" | "reactNativeVersion" | "reactNativeDevVersion"
  >
>;

function minVersion(version: string): string {
  const v = semver.minVersion(version)?.version;
  return v || "0.0.0";
}

export function parseConfig(
  {
    capabilities = [],
    kitType = "library",
    reactNativeVersion = "",
    reactNativeDevVersion,
  }: KitConfig,
  manifest: PackageManifest,
  projectRoot: string
): KitManifest {
  switch (kitType) {
    case "app": {
      const {
        reactNativeVersion,
        capabilities: requiredCapabilities,
      } = getRequirements(manifest, projectRoot);
      requiredCapabilities.push(...capabilities);
      return {
        capabilities: requiredCapabilities,
        kitType,
        reactNativeVersion,
        reactNativeDevVersion:
          reactNativeDevVersion || minVersion(reactNativeVersion),
      };
    }

    case "library": {
      return {
        capabilities,
        kitType,
        reactNativeVersion,
        reactNativeDevVersion:
          reactNativeDevVersion || minVersion(reactNativeVersion),
      };
    }
  }
}
