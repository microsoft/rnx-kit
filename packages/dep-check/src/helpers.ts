import { warn } from "@rnx-kit/console";
import type { PackageManifest } from "@rnx-kit/tools-node/package";
import { writePackage } from "@rnx-kit/tools-node/package";
import detectIndent from "detect-indent";
import fs from "fs";

export function compare<T>(lhs: T, rhs: T): -1 | 0 | 1 {
  if (lhs === rhs) {
    return 0;
  } else if (lhs < rhs) {
    return -1;
  } else {
    return 1;
  }
}

export function concatVersionRanges(versions: string[]): string {
  return "^" + versions.join(" || ^");
}

export function keysOf<T extends Record<string, unknown>>(obj: T): (keyof T)[] {
  return Object.keys(obj);
}

export function modifyManifest(
  pkgPath: string,
  manifest: PackageManifest
): void {
  const content = fs.readFileSync(pkgPath, { encoding: "utf-8" });
  const indent = detectIndent(content).indent || "  ";
  writePackage(pkgPath, manifest, indent);
}

export function omitEmptySections(manifest: PackageManifest): PackageManifest {
  const sections = ["dependencies", "peerDependencies", "devDependencies"];
  for (const sectionName of sections) {
    const section = manifest[sectionName];
    if (typeof section === "object" && Object.keys(section).length === 0) {
      delete manifest[sectionName];
    }
  }
  return manifest;
}

export function printMigrationMessage(): void {
  const banner =
    "⚠️   ⚠️   ⚠️   ⚠️   ⚠️   ⚠️   ⚠️   ⚠️   ⚠️   ⚠️   ⚠️   ⚠️   ⚠️   ⚠️   ⚠️   ⚠️";

  warn(banner);
  warn("'@rnx-kit/dep-check' has been renamed to '@rnx-kit/align-deps'!");
  warn(
    "You can replace '@rnx-kit/dep-check' with '@rnx-kit/align-deps' in your 'package.json' and your configurations will be automatically upgraded."
  );
  warn(
    "For reasoning and more details, you can read the RFC: https://github.com/microsoft/rnx-kit/pull/1757"
  );
  warn(banner);
}
