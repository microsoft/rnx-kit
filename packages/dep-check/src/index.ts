export { capabilitiesFor } from "./capabilities";
export { checkPackageManifest } from "./check";
export { Args, cli } from "./cli";
export { updatePackageManifest } from "./manifest";
export * from "./types";
