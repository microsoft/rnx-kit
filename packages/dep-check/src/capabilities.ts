import type { Capability, KitCapabilities } from "@rnx-kit/config";
import type { PackageManifest } from "@rnx-kit/tools-node/package";
import semver from "semver";
import { getProfilesFor, getProfileVersionsFor } from "./profiles";
import { concatVersionRanges, keysOf } from "./helpers";
import type { CapabilitiesOptions, Package, Profile } from "./types";

export function capabilitiesFor(
  { dependencies, devDependencies, peerDependencies }: PackageManifest,
  { kitType = "library", customProfilesPath }: CapabilitiesOptions = {}
): Partial<KitCapabilities> | undefined {
  const targetReactNativeVersion =
    peerDependencies?.["react-native"] || dependencies?.["react-native"];
  if (!targetReactNativeVersion) {
    return undefined;
  }

  const profiles = getProfilesFor(targetReactNativeVersion, customProfilesPath);
  const packageToCapabilityMap: Record<string, Capability[]> = {};
  profiles.forEach((profile) => {
    keysOf(profile).reduce((result, capability) => {
      const { name } = profile[capability];
      if (!result[name]) {
        result[name] = [capability];
      } else {
        result[name].push(capability);
      }
      return result;
    }, packageToCapabilityMap);
  });

  const reactNativeVersion = concatVersionRanges(
    getProfileVersionsFor(targetReactNativeVersion)
  );

  return {
    reactNativeVersion,
    ...(kitType === "library"
      ? {
          reactNativeDevVersion:
            devDependencies?.["react-native"] ||
            semver.minVersion(reactNativeVersion)?.version,
        }
      : undefined),
    kitType,
    capabilities: Array.from(
      keysOf({
        ...dependencies,
        ...peerDependencies,
        ...devDependencies,
      }).reduce<Set<Capability>>((result, dependency) => {
        if (dependency in packageToCapabilityMap) {
          packageToCapabilityMap[dependency].forEach((capability) => {
            result.add(capability);
          });
        }
        return result;
      }, new Set<Capability>())
    ).sort(),
  };
}

export function resolveCapabilities(
  capabilities: Capability[],
  profiles: Profile[]
): Record<string, Package[]> {
  const unresolvedCapabilities = new Set<string>();
  const packages = capabilities.reduce<Record<string, Package[]>>(
    (dependencies, capability) => {
      profiles.forEach((profile) => {
        const pkg = profile[capability];
        if (!pkg) {
          unresolvedCapabilities.add(capability);
          return;
        }

        const { name, version } = pkg;
        if (name in dependencies) {
          const versions = dependencies[name];
          if (!versions.find((current) => current.version === version)) {
            versions.push(pkg);
          }
        } else {
          dependencies[name] = [pkg];
        }
      });
      return dependencies;
    },
    {}
  );

  if (unresolvedCapabilities.size > 0) {
    const message = Array.from(unresolvedCapabilities).reduce(
      (lines, capability) => (lines += `\n    ${capability}`),
      "The following capabilities could not be resolved for one or more profiles:"
    );

    console.warn(message);
  }

  return packages;
}
