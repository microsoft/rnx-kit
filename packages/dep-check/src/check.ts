import { getKitCapabilities, getKitConfig } from "@rnx-kit/config";
import { error, info, warn } from "@rnx-kit/console";
import { isPackageManifest, readPackage } from "@rnx-kit/tools-node/package";
import chalk from "chalk";
import { diffLinesUnified } from "jest-diff";
import path from "path";
import { getRequirements } from "./dependencies";
import { findBadPackages } from "./findBadPackages";
import { modifyManifest, printMigrationMessage } from "./helpers";
import { updatePackageManifest } from "./manifest";
import { getProfilesFor, resolveCustomProfiles } from "./profiles";
import type { CheckConfig, CheckOptions, Command } from "./types";

export function getCheckConfig(
  manifestPath: string,
  {
    loose,
    uncheckedReturnCode = 0,
    supportedVersions,
    targetVersion,
  }: CheckOptions
): number | CheckConfig {
  const manifest = readPackage(manifestPath);
  if (!isPackageManifest(manifest)) {
    error(
      `'${manifestPath}' does not contain a valid package manifest - please make sure it's not missing 'name' or 'version'`
    );
    return 1;
  }

  const badPackages = findBadPackages(manifest);
  if (badPackages) {
    warn(
      `Known bad packages are found in '${manifest.name}':\n` +
        badPackages
          .map((pkg) => `    ${pkg.name}@${pkg.version}: ${pkg.reason}`)
          .join("\n")
    );
  }

  const projectRoot = path.dirname(manifestPath);
  const kitConfig = getKitConfig({ cwd: projectRoot });
  if (!kitConfig) {
    return uncheckedReturnCode;
  }

  const {
    reactNativeVersion: targetReactNativeVersion,
    reactNativeDevVersion,
    kitType,
    capabilities: targetCapabilities,
    customProfiles,
  } = getKitCapabilities({
    // React Native versions declared in the package's config should always
    // override the ones specified with the `--vigilant` flag.
    ...(supportedVersions
      ? { reactNativeVersion: supportedVersions }
      : undefined),
    // We should not set dev version if the package is configured. It may have
    // been intentionally left out to reuse `reactNativeVersion`.
    ...(!kitConfig.reactNativeVersion && targetVersion
      ? { reactNativeDevVersion: targetVersion }
      : undefined),
    ...kitConfig,
  });

  const customProfilesPath = resolveCustomProfiles(projectRoot, customProfiles);

  const { reactNativeVersion, capabilities: requiredCapabilities } =
    getRequirements(
      targetReactNativeVersion,
      kitType,
      manifest,
      projectRoot,
      customProfilesPath,
      { loose }
    );
  requiredCapabilities.push(...targetCapabilities);

  return {
    kitType,
    reactNativeVersion,
    reactNativeDevVersion,
    capabilities: requiredCapabilities,
    customProfilesPath,
    manifest,
  };
}

export function checkPackageManifest(
  manifestPath: string,
  options: CheckOptions
): number {
  const result = options.config || getCheckConfig(manifestPath, options);
  if (typeof result === "number") {
    return result;
  }

  const {
    kitType,
    reactNativeVersion,
    reactNativeDevVersion,
    capabilities,
    customProfilesPath,
    manifest,
  } = result;

  if (capabilities.length === 0) {
    return options.uncheckedReturnCode || 0;
  }

  const updatedManifest = updatePackageManifest(
    manifest,
    capabilities,
    getProfilesFor(reactNativeVersion, customProfilesPath),
    getProfilesFor(reactNativeDevVersion, customProfilesPath),
    kitType
  );

  // Don't fail when manifests only have whitespace differences.
  const updatedManifestJson = JSON.stringify(updatedManifest, undefined, 2);
  const normalizedManifestJson = JSON.stringify(manifest, undefined, 2);

  if (updatedManifestJson !== normalizedManifestJson) {
    if (options.write) {
      modifyManifest(manifestPath, updatedManifest);
    } else {
      const diff = diffLinesUnified(
        normalizedManifestJson.split("\n"),
        updatedManifestJson.split("\n"),
        {
          aAnnotation: "Current",
          aColor: chalk.red,
          bAnnotation: "Expected",
          bColor: chalk.green,
        }
      );
      console.log(diff);

      error(
        "Changes are needed to satisfy all requirements. Re-run with `--write` to have dep-check apply them."
      );

      const url = chalk.bold("https://aka.ms/dep-check");
      info(`Visit ${url} for more information about dep-check.`);

      printMigrationMessage();

      return 1;
    }
  }

  printMigrationMessage();

  return 0;
}

export function makeCheckCommand(options: CheckOptions): Command {
  return (manifest: string) => {
    return checkPackageManifest(manifest, options);
  };
}
