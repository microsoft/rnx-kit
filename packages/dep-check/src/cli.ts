#!/usr/bin/env node

import type { KitType } from "@rnx-kit/config";
import { error, warn } from "@rnx-kit/console";
import { hasProperty } from "@rnx-kit/tools-language/properties";
import { findPackageDir } from "@rnx-kit/tools-node/package";
import {
  findWorkspacePackages,
  findWorkspaceRoot,
} from "@rnx-kit/tools-workspaces";
import isString from "lodash/isString";
import * as path from "path";
import { makeCheckCommand } from "./check";
import { initializeConfig } from "./initialize";
import { resolveCustomProfiles } from "./profiles";
import { makeSetVersionCommand } from "./setVersion";
import type { Args, Command } from "./types";
import { makeVigilantCommand } from "./vigilant";

function ensureKitType(type: string): KitType | undefined {
  switch (type) {
    case "app":
    case "library":
      return type;
    default:
      return undefined;
  }
}

async function getManifests(
  packages: (string | number)[] | undefined
): Promise<string[] | undefined> {
  if (Array.isArray(packages)) {
    return packages.reduce<string[]>((result, pkg) => {
      const dir = findPackageDir(pkg.toString());
      if (dir) {
        result.push(path.join(dir, "package.json"));
      }
      return result;
    }, []);
  }

  const packageDir = findPackageDir();
  if (!packageDir) {
    return undefined;
  }

  // Make sure we don't return all packages when dep-check is run inside a
  // package that just happened to be part of a workspace.
  const currentPackageJson = path.join(packageDir, "package.json");
  try {
    if ((await findWorkspaceRoot()) !== packageDir) {
      return [currentPackageJson];
    }
  } catch (_) {
    return [currentPackageJson];
  }

  try {
    const allPackages = (await findWorkspacePackages()).map((p) =>
      path.join(p, "package.json")
    );
    allPackages.push(currentPackageJson);
    return allPackages;
  } catch (e) {
    if (hasProperty(e, "message")) {
      error(e.message);
      return undefined;
    }

    throw e;
  }
}

function makeInitializeCommand(
  kitType: string,
  customProfiles: string | undefined
): Command | undefined {
  const verifiedKitType = ensureKitType(kitType);
  if (!verifiedKitType) {
    error(`Invalid kit type: '${kitType}'`);
    return undefined;
  }

  return (manifest: string) => {
    initializeConfig(manifest, {
      kitType: verifiedKitType,
      customProfilesPath: customProfiles,
    });
    return 0;
  };
}

function reportConflicts(conflicts: [string, string][], args: Args): boolean {
  return conflicts.reduce<boolean>((result, [lhs, rhs]) => {
    if (lhs in args && rhs in args) {
      error(`--${lhs} and --${rhs} cannot both be specified at the same time.`);
      return true;
    }
    return result;
  }, false);
}

async function makeCommand(args: Args): Promise<Command | undefined> {
  const conflicts: [string, string][] = [
    ["init", "vigilant"],
    ["init", args.write ? "write" : "no-write"],
    ["set-version", args.write ? "write" : "no-write"],
  ];
  if (reportConflicts(conflicts, args)) {
    return undefined;
  }

  const {
    "custom-profiles": customProfiles,
    "exclude-packages": excludePackages,
    init,
    loose,
    "set-version": setVersion,
    vigilant,
    write,
  } = args;

  if (isString(init)) {
    return makeInitializeCommand(init, customProfiles?.toString());
  }

  // When `--set-version` is without a value, `setVersion` is an empty string if
  // invoked directly. When invoked via `@react-native-community/cli`,
  // `setVersion` is `true` instead.
  if (setVersion || isString(setVersion)) {
    return makeSetVersionCommand(setVersion);
  }

  if (isString(vigilant)) {
    const customProfilesPath = resolveCustomProfiles(
      process.cwd(),
      customProfiles?.toString()
    );
    return makeVigilantCommand({
      customProfiles: customProfilesPath,
      excludePackages: excludePackages?.toString(),
      loose,
      versions: vigilant.toString(),
      write,
    });
  }

  return makeCheckCommand({ loose, write });
}

export async function cli({ packages, ...args }: Args): Promise<void> {
  const command = await makeCommand(args);
  if (!command) {
    process.exit(1);
  }

  const manifests = await getManifests(packages);
  if (!manifests) {
    error("Could not find package root");
    process.exit(1);
  }

  // We will optimistically run through all packages regardless of failures. In
  // most scenarios, this should be fine: Both init and check+write write to
  // disk only when everything is in order for the target package. Packages with
  // invalid or missing configurations are skipped.
  const exitCode = manifests.reduce((exitCode: number, manifest: string) => {
    try {
      return command(manifest) || exitCode;
    } catch (e) {
      if (hasProperty(e, "message")) {
        const currentPackageJson = path.relative(process.cwd(), manifest);

        if (hasProperty(e, "code") && e.code === "ENOENT") {
          warn(`${currentPackageJson}: ${e.message}`);
          return exitCode;
        }

        error(`${currentPackageJson}: ${e.message}`);
        return exitCode || 1;
      }

      throw e;
    }
  }, 0);

  process.exit(exitCode);
}

if (require.main === module) {
  require("yargs").usage(
    "$0 [packages...]",
    "Dependency checker for React Native apps",
    {
      "custom-profiles": {
        description:
          "Path to custom profiles. This can be a path to a JSON file, a `.js` file, or a module name.",
        type: "string",
        requiresArg: true,
      },
      "exclude-packages": {
        description:
          "Comma-separated list of package names to exclude from inspection.",
        type: "string",
        requiresArg: true,
        implies: "vigilant",
      },
      init: {
        description:
          "Writes an initial kit config to the specified 'package.json'.",
        choices: ["app", "library"],
        conflicts: ["vigilant"],
      },
      loose: {
        default: false,
        description:
          "Determines how strict the React Native version requirement should be. Useful for apps that depend on a newer React Native version than their dependencies declare support for.",
        type: "boolean",
      },
      "set-version": {
        description:
          "Sets `reactNativeVersion` and `reactNativeDevVersion` for any configured package. There is an interactive prompt if no value is provided. The value should be a comma-separated list of `react-native` versions to set, where the first number specifies the development version. Example: `0.64,0.63`",
        type: "string",
        conflicts: ["init", "vigilant"],
      },
      vigilant: {
        description:
          "Inspects packages regardless of whether they've been configured. Specify a comma-separated list of profile versions to compare against, e.g. `0.63,0.64`. The first number specifies the target version.",
        type: "string",
        requiresArg: true,
        conflicts: ["init"],
      },
      write: {
        default: false,
        description: "Writes changes to the specified 'package.json'.",
        type: "boolean",
      },
    },
    cli
  ).argv;
}
