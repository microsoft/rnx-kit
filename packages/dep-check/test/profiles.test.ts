import semver from "semver";
import {
  defaultProfiles,
  getProfilesFor,
  getProfileVersionsFor,
  profilesSatisfying,
} from "../src/profiles";
import profile_0_62 from "../src/profiles/profile-0.62";
import profile_0_63 from "../src/profiles/profile-0.63";
import profile_0_64 from "../src/profiles/profile-0.64";
import { ProfileVersion } from "../src/types";
import { mockResolver } from "./helpers";

describe("defaultProfiles", () => {
  test("matches react-native versions", () => {
    const includePrerelease = {
      includePrerelease: true,
    };
    Object.entries(defaultProfiles).forEach(([version, capabilities]) => {
      const versionRange = "^" + version;
      Object.entries(capabilities).forEach(([capability, pkg]) => {
        if (capability === "core") {
          expect(
            "version" in pkg &&
              semver.subset(pkg.version, versionRange, includePrerelease)
          ).toBe(true);
        }
      });
    });
  });
});

describe("getProfileVersionsFor()", () => {
  test("returns profile versions for specific version", () => {
    expect(getProfileVersionsFor("0.61.5")).toEqual(["0.61"]);
    expect(getProfileVersionsFor("0.62.2")).toEqual(["0.62"]);
    expect(getProfileVersionsFor("0.63.4")).toEqual(["0.63"]);
    expect(getProfileVersionsFor("0.64.0")).toEqual(["0.64"]);
  });

  test("returns profile for one version range", () => {
    expect(getProfileVersionsFor("^0.62.2")).toEqual(["0.62"]);
    expect(getProfileVersionsFor("^0.63.4")).toEqual(["0.63"]);
    expect(getProfileVersionsFor("^0.64.0")).toEqual(["0.64"]);
  });

  test("returns profiles for bigger version ranges", () => {
    const profiles = getProfileVersionsFor(">=0.62.2");
    expect(profiles).toEqual(["0.62", "0.63", "0.64", "0.65"]);
  });

  test("returns profiles for multiple version ranges", () => {
    const profiles = getProfileVersionsFor("^0.63.0 || ^0.64.0");
    expect(profiles).toEqual(["0.63", "0.64"]);
  });

  test("returns an empty array when an unsupported version is provided", () => {
    expect(getProfileVersionsFor("^0.60.6")).toEqual([]);
  });

  test("throws when an invalid version is provided", () => {
    expect(() => getProfileVersionsFor("invalid")).toThrowError(
      "Invalid 'react-native' version"
    );
  });
});

describe("getProfilesFor()", () => {
  const consoleErrorSpy = jest.spyOn(global.console, "error");

  beforeEach(() => {
    consoleErrorSpy.mockReset();
  });

  afterAll(() => {
    jest.clearAllMocks();
  });

  test("returns profile for specific version", () => {
    const profiles = getProfilesFor("0.64.0", undefined);
    expect(profiles).toEqual([profile_0_64]);
    expect(consoleErrorSpy).not.toBeCalled();
  });

  test("returns profile for one version range", () => {
    expect(getProfilesFor("^0.62.2", undefined)).toEqual([profile_0_62]);
    expect(getProfilesFor("^0.63.4", undefined)).toEqual([profile_0_63]);
    expect(getProfilesFor("^0.64.0", undefined)).toEqual([profile_0_64]);
  });

  test("returns profiles for bigger version ranges", () => {
    const profiles = getProfilesFor(">=0.62.2", undefined);
    expect(profiles.slice(0, 3)).toEqual([
      profile_0_62,
      profile_0_63,
      profile_0_64,
    ]);
    expect(consoleErrorSpy).not.toBeCalled();
  });

  test("returns profiles for multiple version ranges", () => {
    const profiles = getProfilesFor("^0.63.0 || ^0.64.0", undefined);
    expect(profiles).toEqual([profile_0_63, profile_0_64]);
    expect(consoleErrorSpy).not.toBeCalled();
  });

  test("throws when an unsupported version is provided", () => {
    expect(() => getProfilesFor("^0.60.6", undefined)).toThrowError(
      "Unsupported 'react-native' version"
    );
    expect(consoleErrorSpy).not.toBeCalled();
  });

  test("throws when an invalid version is provided", () => {
    expect(() => getProfilesFor("invalid", undefined)).toThrowError(
      "Invalid 'react-native' version"
    );
    expect(consoleErrorSpy).not.toBeCalled();
  });

  test("throws if custom profiles module does not exist", () => {
    expect(() =>
      getProfilesFor("^0.59", "non-existent-profiles-module")
    ).toThrow(`Cannot find module 'non-existent-profiles-module'`);

    expect(consoleErrorSpy).toBeCalledTimes(2);
  });

  test("throws if custom profiles module path is not returned", () => {
    expect(() =>
      getProfilesFor("^0.59", "undefined-profiles-module", {
        moduleResolver: mockResolver(),
      })
    ).toThrow(`Cannot find module 'undefined-profiles-module'`);

    expect(consoleErrorSpy).toBeCalledTimes(1);
  });

  test("throws if custom profiles module does not default export an object", () => {
    jest.mock("bad-profiles-module", () => null, { virtual: true });

    expect(() =>
      getProfilesFor("^0.59", "bad-profiles-module", {
        moduleResolver: mockResolver("bad-profiles-module"),
      })
    ).toThrow("'bad-profiles-module' doesn't default export profiles");

    expect(consoleErrorSpy).toBeCalledTimes(1);
  });

  test("appends custom profiles", () => {
    const skynet = { name: "skynet", version: "1.0.0" };
    jest.mock(
      "good-profiles-module",
      () => ({ "0.62": { [skynet.name]: skynet } }),
      { virtual: true }
    );

    const [profile_0_62, profile_0_63] = getProfilesFor(
      "^0.62 || ^0.63",
      "good-profiles-module",
      { moduleResolver: mockResolver("good-profiles-module") }
    );

    expect(skynet.name in profile_0_62).toBe(true);
    expect(skynet.name in profile_0_63).toBe(false);

    expect(consoleErrorSpy).not.toBeCalled();
  });
});

describe("profilesSatisfying()", () => {
  const profileVersions: ProfileVersion[] = ["0.61", "0.62", "0.63", "0.64"];

  test("returns an empty array when no profiles can satisfy requirements", () => {
    expect(profilesSatisfying([], "^0.63.0 || ^0.64.0")).toEqual([]);
    expect(profilesSatisfying(["0.61"], "^0.63.0 || ^0.64.0")).toEqual([]);
  });

  test("returns all profiles satisfying version", () => {
    expect(profilesSatisfying(profileVersions, "0.63.2")).toEqual(["0.63"]);
    expect(profilesSatisfying(profileVersions, "0.64.0-rc1")).toEqual(["0.64"]);
    expect(profilesSatisfying(profileVersions, "0.64.0")).toEqual(["0.64"]);
  });

  test("returns all profiles satisfying version range", () => {
    expect(
      profilesSatisfying(profileVersions, "^0.63.0-0 || ^0.64.0-0")
    ).toEqual(["0.63", "0.64"]);
    expect(profilesSatisfying(profileVersions, "^0.63.0 || ^0.64.0")).toEqual([
      "0.63",
      "0.64",
    ]);
  });
});
