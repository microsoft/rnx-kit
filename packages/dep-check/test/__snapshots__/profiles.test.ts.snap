// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`loadCustomProfiles() loads valid custom profiles 1`] = `
Object {
  "0.65": Object {
    "format": Object {
      "devOnly": true,
      "name": "prettier",
      "version": "^2.5.1",
    },
  },
  "0.66": Object {
    "format": Object {
      "devOnly": true,
      "name": "prettier",
      "version": "^2.5.1",
    },
  },
  "0.67": Object {
    "format": Object {
      "devOnly": true,
      "name": "prettier",
      "version": "^2.5.1",
    },
  },
}
`;

exports[`loadCustomProfiles() prepends root-level capabilities to all profiles 1`] = `
Object {
  "0.61": Object {
    "format": Object {
      "devOnly": true,
      "name": "prettier",
      "version": "^2.5.1",
    },
  },
  "0.62": Object {
    "format": Object {
      "devOnly": true,
      "name": "prettier",
      "version": "^2.5.1",
    },
  },
  "0.63": Object {
    "format": Object {
      "devOnly": true,
      "name": "prettier",
      "version": "^2.5.1",
    },
  },
  "0.64": Object {
    "format": Object {
      "devOnly": true,
      "name": "prettier",
      "version": "^2.5.1",
    },
  },
  "0.65": Object {
    "format": Object {
      "devOnly": true,
      "name": "prettier",
      "version": "^2.5.1",
    },
  },
  "0.66": Object {
    "format": Object {
      "devOnly": true,
      "name": "prettier",
      "version": "^2.5.1",
    },
    "test": Object {
      "devOnly": true,
      "name": "jest",
      "version": "26.0",
    },
  },
  "0.67": Object {
    "format": Object {
      "devOnly": true,
      "name": "prettier",
      "version": "3.0",
    },
    "test": Object {
      "devOnly": true,
      "name": "jest",
      "version": "27.0",
    },
  },
  "0.68": Object {
    "format": Object {
      "devOnly": true,
      "name": "prettier",
      "version": "^2.5.1",
    },
  },
  "0.69": Object {
    "format": Object {
      "devOnly": true,
      "name": "prettier",
      "version": "^2.5.1",
    },
  },
  "0.70": Object {
    "format": Object {
      "devOnly": true,
      "name": "prettier",
      "version": "^2.5.1",
    },
  },
}
`;
